src/relay/ir/expr.cc

```diff
@@ -71,6 +71,29 @@ Tuple::Tuple(tvm::Array<relay::Expr> fields, Span span) {
   data_ = std::move(n);
 }
 
+Tuple Tuple::WithFields(Optional<Array<Expr>> opt_fields, Optional<Span> opt_span) {
+  Array<Expr> fields = opt_fields.value_or(get()->fields);
```

Comment:
you want to make it as a global function https://github.com/apache/tvm/blob/main/include/tvm/ir/attrs.h#L371, so that you do std::move and COW when there is a single copy.

Reply:
To elaborate, `x.WithFields` in the current case might mutate x if it is the only copy. 

```c++
y = x.WithFields(..., ...)
// if x is the only copy, it will change x itself. Which may not be the intention of the API.
```

instead if we do 
```c++
y = WithFields(std::move(x), .., ...)_
```
and x is the only copy, we know for certainly that x will not be used in the future and it is safe to re-use the memory space

Reply:
The intention is mutating if x is the only copy.

Reply:
I see, it might be worth thinking a bit about the tradeoff since the original semantics on the python and c++ side is WithAttr do not mutate the original value that is passed in(unless it is moved in and ensures that there is no followup).

Since a common usage pattern may ends up look like this:

```python
x = PrimFunc(args, body)
# y is a new copy that comes with a new attr
# x remains as the original case
y = x.with_attr("new_attr", value)

record([x, y])
``` 

If it is a mution semantics, then we might want to rename it as SetXYZ, like Array.Set

Reply:
OK, thanks for the feedback. I just changed the signature to be similar to WithAttrs and I removed the COW functionality. Now it will return the argument tuple if all the fields passed in are the same, otherwise it creates a copy and returns that.